proposals/NNNN-package-access-modifier.md

```diff
@@ -0,0 +1,226 @@
+
+We want to follow the optimization model for `public`; a `public` class has no subclasses outside of its defining module, and a `public` class member has no overrides outside of its defining module. The optimizer can take advantage of this information in whole-module builds to replace dynamic dispatch with static dispatch among other things. 
+
+## Future Directions
+
+Subclassing
+* A separate access modifier will likely need to be introduced to allow subclassing and overriding within a package. The reason for a new access modifier is described in the Details section.
+
+Namespacing 
+* A package name might potentially be added to a mangled name. This could help disambiguate module names that clash.
+
+Optimizations
+* A package containing several modules can be treated as a resilience domain. If same-package clients need access to module binaries, they don't need to be independently rebuildable and could have an unstable ABI; they could avoid resilience overhead and unnecessary language rules. 
+
+* By default package symbols are exported in the final libraries/executables. We plan to introduce a build setting that allows users to decide whether to hide package symbols for staticaly linked libraries. Enabling package symbols to be hidden would help with size optimizations.
+
+## Source compatibility
+
+A new keyword `package` is added as a new access modifier. It is a contextual keyword and is an additive change. Symbols that are named `package` should not require renaming, and the source code as is should continue to work. 
+
+## Effect on ABI stability
+
+Boundaries between separately-built modules within a package are still potentially ABI boundaries. The ABI for package symbols is not different from the ABI for public symbols, although in the future we plan to add an option to not export package symbols that can be resolved within an image. 
+
+## Alternatives considered
+
+A current workaround for the scenario in the Motivation is to use `@_spi` or `@_implemenationOnly`. Each option has caveats. The `@_spi` requires a group name, which makes it verbose to use and harder to keep track of, and `@_implementationOnly` can be too limiting as we want to be able to restrict access to only portions of APIs. There are also hacky workarounds such as `@testable` and the `-disable-access-control` flag. They elevate all `internal` (and `private` with the flag) symbols to `public`. These options are unstable and also quickly lead to an increase of the binary and the shared cache size, not to mention symbol name clashes.
+
+Instead of a package, submodules could have been considered. Introducing another "umbrella" module or dividing a module into "sub" modules, however, would require much more significant work and would be both source and ABI breaking. A package already exists as a unit corresponding to a group of modules, and treating it as a unit above a module in the grouping hierarchy enables a nautral transition into a submodule-like structure. 
```

Comment:
I think you both want to talk about what submodules mean and argue against why they don't really solve the problem.  ("We already have packages" is not a good argument.  If submodules were a better solution, then we should add submodules, not just try to pick low-hanging fruit when it leads us to a worse solution.)  Maybe something like this?

Instead of adding a new `package` access level above modules, we could allow modules to contain other modules as components.  This is an idea often called "submodules".  Packages would then define an "umbrella" module that contains the package's modules as components. However, there are several weaknesses in this approach:
- It doesn't actually solve the problem by itself. Submodule APIs would still need to be able to declare whether they're usable outside of the umbrella or not, and that would require an access modifier. It might be written in a more general way, like `internal(MyPackage)`, but that generality would also make it more verbose.
- Submodule structure would be part of the source language, so it would naturally be source- and ABI-affecting.  For example, programmers could use the parent module's name to qualify identifiers, and symbols exported by a submodule would include the parent module's name.  This means that splitting a module into submodules or adding an umbrella parent module would be much more impactful than desired; ideally, those changes would be purely internal and not change a module's public interface.  It also means that these changes would end up permanently encoding package structure.
- The "umbrella" submodule structure doesn't work for all packages.  Some packages include multiple "top-level" modules which share common dependencies.  Forcing these to share a common umbrella in order to use package-private dependencies is not desirable.
- In a few cases, the ABI and source impact above *would* be desirable.  For example, many packages contain internal `Utility` modules; if these were declared as submodules, they would naturally be namespaced to the containing package, eliminating spurious collisions. However, such modules are generally not meant to be usable outside of the package at all. It is a reasonable future direction to allow whole modules to be made package-private, which would also make it reasonable to automatically namespace them.